File: sample-app/src/main/java/dev/oneuiproject/oneuiexample/ui/fragment/icons/adapter/IconsAdapter.java
package dev.oneuiproject.oneuiexample.ui.fragment.icons.adapter;

import android.content.Context;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.CheckBox;
import android.widget.ImageView;
import android.widget.SectionIndexer;
import android.widget.TextView;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.collection.ScatterSet;
import androidx.recyclerview.widget.RecyclerView;

import com.sec.sesl.tester.R;

import java.util.ArrayList;
import java.util.List;

import dev.oneuiproject.oneui.delegates.AllSelectorState;
import dev.oneuiproject.oneui.delegates.MultiSelector;
import dev.oneuiproject.oneui.delegates.MultiSelectorDelegate;
import dev.oneuiproject.oneui.delegates.SectionIndexerDelegate;
import dev.oneuiproject.oneui.utils.SearchHighlighter;
import kotlin.Unit;
import kotlin.jvm.functions.Function1;

public class IconsAdapter extends RecyclerView.Adapter<IconsAdapter.ViewHolder>
        implements SectionIndexer, MultiSelector<Integer> {

    private final List<Integer> mIconsId = new ArrayList<>();
    private final List<Integer> filteredIconsId = new ArrayList<>();  // For filtered data
    private final Context mContext;

    private final MultiSelectorDelegate<Integer> mSelectionDelegate = new MultiSelectorDelegate<>();
    private final SectionIndexerDelegate<Integer> indexerDelegate;

    private final SearchHighlighter highlighter = new SearchHighlighter();
    private String query = "";

    public interface OnItemClickListener {
        void onItemClick(Integer iconId, int position);
        void onItemLongClick(Integer iconId, int position);
    }

    public OnItemClickListener onItemClickListener = null;

    public IconsAdapter(Context context) {
        mContext = context;
        indexerDelegate = new SectionIndexerDelegate<>(context, this::getIndexerLabel);
    }

    private String getIndexerLabel(Integer iconId) {
        char firstChar = mContext.getResources().getResourceEntryName(iconId)
                .replace("ic_oui_", "").charAt(0);
        if (Character.isDigit(firstChar)) {
            return "#";
        }
        return Character.toString(firstChar).toUpperCase();
    }

    public void submitList(List<Integer> list) {
        mIconsId.clear();
        mIconsId.addAll(list);
        filteredIconsId.clear();
        filteredIconsId.addAll(list);  // Initialize filtered list with all items
        notifyDataSetChanged();
        indexerDelegate.updateSections(list, false);
        updateSelectableIds(filteredIconsId);
    }

    // Filtering method
    public void filter(String query) {
        if (this.query.equals(query)) return;
        this.query = query;

        filteredIconsId.clear();
        if (query == null || query.isEmpty()) {
            filteredIconsId.addAll(mIconsId);  // No filter applied, show all items
        } else {
            String lowerCaseQuery = query.toLowerCase();
            for (Integer iconId : mIconsId) {
                String iconName = mContext.getResources().getResourceEntryName(iconId).toLowerCase();
                if (iconName.contains(lowerCaseQuery)) {
                    filteredIconsId.add(iconId);  // Add matching items to filtered list
                }
            }
        }
        notifyDataSetChanged();
        indexerDelegate.updateSections(filteredIconsId, false);
    }

    public int getItem(int position) {
        return filteredIconsId.get(position);
    }

    @Override
    public int getItemCount() {
        return filteredIconsId.size();
    }

    @NonNull
    @Override
    public IconsAdapter.ViewHolder onCreateViewHolder(@NonNull ViewGroup parent, int viewType) {
        LayoutInflater inflater = LayoutInflater.from(mContext);
        View view = inflater.inflate(R.layout.sample3_view_icon_listview_item, parent, false);
        ViewHolder vh = new ViewHolder(view);
        vh.itemView.setOnClickListener(v -> {
            if (onItemClickListener != null) {
                onItemClickListener.onItemClick(filteredIconsId.get(vh.getBindingAdapterPosition()),
                        vh.getBindingAdapterPosition());
            }
        });
        vh.itemView.setOnLongClickListener(v -> {
            if (onItemClickListener != null) {
                onItemClickListener.onItemLongClick(filteredIconsId.get(vh.getBindingAdapterPosition()),
                        vh.getBindingAdapterPosition());
            }
            return true;
        });
        return vh;
    }

    @Override
    public void onBindViewHolder(ViewHolder holder, int position) {
        Integer iconId = filteredIconsId.get(position);
        holder.imageView.setImageResource(iconId);
        holder.textView.setText(highlighter.invoke(mContext.getResources().getResourceEntryName(iconId), query));
        holder.checkBox.setVisibility(isActionMode() ? View.VISIBLE : View.GONE);
        holder.checkBox.setChecked(isSelected(iconId));
    }

    // Implement SectionIndexer methods
    @Override
    public Object[] getSections() {
        return indexerDelegate.getSections();
    }

    @Override
    public int getPositionForSection(int sectionIndex) {
        return indexerDelegate.getPositionForSection(sectionIndex);
    }

    @Override
    public int getSectionForPosition(int position) {
        return indexerDelegate.getSectionForPosition(position);
    }

    // MultiSelector interface methods (delegate implementation)
    @Override
    public void configure(@NonNull RecyclerView recyclerView,
                          @Nullable Object selectionChangePayload,
                          @Nullable Function1<? super Integer, ? extends Integer> selectionId,
                          @NonNull Function1<? super AllSelectorState, Unit> allSelectorState) {
        mSelectionDelegate.configure(recyclerView, selectionChangePayload, selectionId, allSelectorState);
    }

    @Override
    public void updateSelectableIds(@NonNull List<? extends Integer> selectableIds) {
        mSelectionDelegate.updateSelectableIds(selectableIds);
    }

    @Override
    public void onToggleActionMode(boolean isActionMode, @Nullable Integer[] initialSelectedIds) {
        mSelectionDelegate.onToggleActionMode(isActionMode, initialSelectedIds);
    }

    @Override
    public void onToggleItem(Integer selectionId, int position) {
        mSelectionDelegate.onToggleItem(selectionId, position);
    }

    @Override
    public void onToggleSelectAll(boolean isSelectAll) {
        mSelectionDelegate.onToggleSelectAll(isSelectAll);
    }

    @Override
    public boolean isSelected(Integer selectionId) {
        return mSelectionDelegate.isSelected(selectionId);
    }

    @NonNull
    @Override
    public ScatterSet<Integer> getSelectedIds() {
        return mSelectionDelegate.getSelectedIds();
    }

    @Override
    public boolean isActionMode() {
        return mSelectionDelegate.isActionMode();
    }

    // ViewHolder class
    public static class ViewHolder extends RecyclerView.ViewHolder {
        ImageView imageView;
        TextView textView;
        CheckBox checkBox;

        ViewHolder(View itemView) {
            super(itemView);
            imageView = itemView.findViewById(R.id.icon_list_item_icon);
            textView = itemView.findViewById(R.id.icon_list_item_text);
            checkBox = itemView.findViewById(R.id.icon_list_item_checkbox);
        }
    }
}

File: sample-app/src/main/java/dev/oneuiproject/oneuiexample/ui/main/fragments/recyclerview/icons/adapter/IconsAdapter.kt
package dev.oneuiproject.oneuiexample.ui.main.fragments.recyclerview.icons.adapter

import android.view.LayoutInflater
import android.view.ViewGroup
import android.widget.ImageView
import android.widget.TextView
import androidx.recyclerview.widget.DiffUtil
import androidx.recyclerview.widget.RecyclerView
import com.sec.sesl.tester.databinding.ViewIconListviewItemBinding
import dev.oneuiproject.oneui.layout.ToolbarLayout.AllSelectorState
import dev.oneuiproject.oneui.recyclerview.adapter.IndexedSelectableListAdapter
import dev.oneuiproject.oneui.recyclerview.model.AdapterItem
import dev.oneuiproject.oneui.utils.SearchHighlighter
import dev.oneuiproject.oneui.widget.SelectableLinearLayout
import dev.oneuiproject.oneuiexample.ui.main.fragments.recyclerview.icons.IconListItemUiModel
import dev.oneuiproject.oneuiexample.ui.main.fragments.recyclerview.stargazers.adapter.StargazersAdapter

class IconsAdapter(
    onAllSelectorStateChanged: ((AllSelectorState) -> Unit),
    onBlockActionMode: (() -> Unit)
):
    IndexedSelectableListAdapter<IconListItemUiModel, IconsAdapter.ViewHolder, Int>(
        indexLabelExtractor,
        onAllSelectorStateChanged,
        onBlockActionMode,
        selectableIdsProvider,
        selectionIdProvider,
        null,
        null,
        diffCallback
) {

    var onClickItem: ((IconListItemUiModel, Int) -> Unit)? = null
    var onLongClickItem: (() -> Unit)? = null

    var highlightWord: String = ""
        set(value) {
            if (value != field) {
                field = value
                notifyItemRangeChanged(0, itemCount, StargazersAdapter.Payload.HIGHLIGHT)
            }
        }

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): ViewHolder {
        val view =
            ViewIconListviewItemBinding.inflate(LayoutInflater.from(parent.context), parent, false)
        return ViewHolder(view).apply {
            itemView.apply {
                setOnClickListener {
                    bindingAdapterPosition.let {
                        onClickItem?.invoke(currentList[it], it)
                    }
                }
                setOnLongClickListener {
                    onLongClickItem?.invoke()
                    true
                }
            }
        }
    }

    override fun onBindViewHolder(
        holder: ViewHolder,
        position: Int
    ) = holder.bind(currentList[position])

    override fun onBindViewHolder(holder: ViewHolder, position: Int, payloads: MutableList<Any>) {
        if (payloads.isEmpty()) {
            onBindViewHolder(holder, position)
        } else {
            val item = currentList[position]
            for (payload in payloads.toSet()) {
                when (payload) {
                    StargazersAdapter.Payload.SELECTION_MODE -> holder.bindSelected(item)
                    StargazersAdapter.Payload.HIGHLIGHT -> holder.bindText(item.name)
                }
            }
        }
    }

    inner class ViewHolder internal constructor(binding: ViewIconListviewItemBinding) :
        RecyclerView.ViewHolder(binding.root) {
        private val stringHighlight = SearchHighlighter(binding.root.context)//needs theme context
        private var imageView: ImageView = binding.iconListItemIcon
        private var textView: TextView = binding.iconListItemText

        fun bind(item: IconListItemUiModel) {
            bindSelected(item)
            bindText(item.name)
            imageView.setImageResource(item.id)
            (itemView as? SelectableLinearLayout)?.apply {
                isSelectionMode = isActionMode
                setSelected(isSelected(item.id))
            }
        }

        fun bindText(itemName: String) {
            textView.text = stringHighlight(itemName, highlightWord)
        }

        fun bindSelected(item: IconListItemUiModel) {
            itemView.setSelected(isSelected(item.id))
        }
    }

    companion object {
        private val indexLabelExtractor: (IconListItemUiModel) -> CharSequence =
            { item ->
                val firstChar = item.name.replace("ic_oui_", "")[0]
                if (Character.isDigit(firstChar)) {
                    "#"
                } else {
                    firstChar.toString().uppercase()
                }
            }

        private val selectableIdsProvider: (currentList: List<IconListItemUiModel>) -> List<Int> =
            { list -> list.map { it.id } }

        private val diffCallback = object : DiffUtil.ItemCallback<IconListItemUiModel>() {
            override fun areItemsTheSame(oldItem: IconListItemUiModel, newItem: IconListItemUiModel) = oldItem.id == newItem.id
            override fun areContentsTheSame(oldItem: IconListItemUiModel, newItem: IconListItemUiModel) = oldItem.name == newItem.name
        }

        private val selectionIdProvider: ((RecyclerView, AdapterItem) -> Int) =
            { rv, item -> (rv.adapter as IconsAdapter).getItem(item.position).id }

    }
}
